x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PushDownAndCombineLimits.java

```diff
@@ -102,8 +102,11 @@ private static Limit combineLimits(Limit upper, Limit lower, FoldContext ctx) {
         var upperLimitValue = (int) upper.limit().fold(ctx);
         var lowerLimitValue = (int) lower.limit().fold(ctx);
         // We want to preserve the duplicated() value of the smaller limit.
-        if (lowerLimitValue <= upperLimitValue) {
-            return lower.withLocal(lower.local());
+        if (lowerLimitValue < upperLimitValue) {
+            return lower;
+        } else if (lowerLimitValue == upperLimitValue) {
+            // If any of them is local, we want the local limit
+            return lower.local() ? lower : lower.withLocal(upper.local());
```

Comment:
Why is this correct?

`local == true` means "I got duplicated from an optimization and am thus optional". But `local == false` is generally just a plain `limit` - just because a downstream optimization created another limit that's optional to execute, doesn't mean that we can throw away upstream limits that may well be coming from the user and we need to pipeline break on them?